File: backend/src/plugins/core/manifest.py
"""Parse and validate ``plugin.md`` frontmatter.

The manifest is the single source of truth for a plugin's exported tools,
optional setup script, and optional in-sandbox runtime. See
``docs/architecture/plugins-refactor.md`` §1.1 for the contract.

Path resolution is strict: every declared path must resolve under the plugin
directory (no ``..`` escape) and exist on disk. Tool *binding* (importing the
module and verifying it exposes a ``BaseTool``) is the loader's job
(:mod:`plugins.core.loader`); this module only validates structure.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

__all__ = [
    "ALLOWED_PLUGIN_KINDS",
    "PluginManifest",
    "PluginManifestError",
    "ToolEntry",
    "parse_plugin_manifest",
]


class PluginManifestError(ValueError):
    """Raised when a ``plugin.md`` fails schema validation."""


_NAME_PATTERN = re.compile(r"^[a-z][a-z0-9_]*$")
_FRONTMATTER_RE = re.compile(
    r"\A---\s*\n(?P<frontmatter>.*?)\n---\s*(?:\n(?P<body>.*))?\Z",
    re.DOTALL,
)

# V3 README §Requirement 2 enum — values for ``plugin_kind`` in the audit
# payload. Closer D (Phase 2.6): manifest authors may declare ``kind`` so
# the plugin shim can stamp the real value on ``plugin.*`` events instead of
# defaulting to ``"custom"``.
ALLOWED_PLUGIN_KINDS: frozenset[str] = frozenset(
    {
        "language_server",
        "formatter",
        "indexer",
        "build_daemon",
        "mcp_bridge",
        "custom",
    }
)


@dataclass(frozen=True)
class ToolEntry:
    """One declared tool in a manifest."""

    name: str
    module: Path  # absolute path under the plugin directory


@dataclass(frozen=True)
class PluginManifest:
    """Parsed and validated ``plugin.md``."""

    name: str
    description: str
    tools: tuple[ToolEntry, ...]
    setup: Path | None  # absolute path or None
    runtime: Path | None  # absolute path or None
    source_dir: Path  # absolute path to the plugin directory
    body: str  # markdown body after the frontmatter (informational)
    kind: str | None = None  # one of ALLOWED_PLUGIN_KINDS, or None for unset


def parse_plugin_manifest(plugin_dir: Path) -> PluginManifest:
    """Parse ``<plugin_dir>/plugin.md`` and validate the schema."""
    plugin_dir = plugin_dir.resolve()
    manifest_path = plugin_dir / "plugin.md"
    if not manifest_path.is_file():
        raise PluginManifestError(f"plugin.md missing under {plugin_dir}")

    text = manifest_path.read_text(encoding="utf-8")
    match = _FRONTMATTER_RE.match(text)
    if match is None:
        raise PluginManifestError(
            "plugin.md must begin with a `---`-delimited YAML frontmatter "
            f"block: {manifest_path}"
        )

    try:
        data = yaml.safe_load(match.group("frontmatter")) or {}
    except yaml.YAMLError as exc:
        raise PluginManifestError(
            f"plugin.md frontmatter is not valid YAML: {manifest_path}: {exc}"
        ) from exc
    if not isinstance(data, dict):
        raise PluginManifestError(
            f"plugin.md frontmatter must be a YAML mapping: {manifest_path}"
        )

    name = _require_str(data, "name", manifest_path)
    if not _NAME_PATTERN.match(name):
        raise PluginManifestError(
            f"plugin.md name {name!r} must match {_NAME_PATTERN.pattern}: "
            f"{manifest_path}"
        )
    if name != plugin_dir.name:
        raise PluginManifestError(
            f"plugin.md name {name!r} does not match directory name "
            f"{plugin_dir.name!r}: {manifest_path}"
        )

    description = _require_str(data, "description", manifest_path)
    tools = _parse_tools(data, plugin_dir, manifest_path, plugin_name=name)
    setup = _resolve_setup(data, plugin_dir, manifest_path)
    runtime = _resolve_optional_path(
        data, "runtime", plugin_dir, manifest_path
    )
    kind = _parse_kind(data, manifest_path)

    body_match = match.group("body")
    body = (body_match or "").strip()

    return PluginManifest(
        name=name,
        description=description,
        tools=tuple(tools),
        setup=setup,
        runtime=runtime,
        source_dir=plugin_dir,
        body=body,
        kind=kind,
    )


def _parse_kind(
    data: dict[str, Any],
    manifest_path: Path,
) -> str | None:
    """Validate the optional ``kind`` field against the V3 enum.

    Closer D (Phase 2.6): authors declare one of
    :data:`ALLOWED_PLUGIN_KINDS` so the audit shim stamps the right value
    on ``plugin.*`` events. Unknown values are a hard error — silent typos
    would broaden the schema invisibly.
    """
    raw = data.get("kind")
    if raw is None:
        return None
    if not isinstance(raw, str) or not raw.strip():
        raise PluginManifestError(
            f"plugin.md kind must be a non-empty string when set: {manifest_path}"
        )
    value = raw.strip()
    if value not in ALLOWED_PLUGIN_KINDS:
        allowed = ", ".join(sorted(ALLOWED_PLUGIN_KINDS))
        raise PluginManifestError(
            f"plugin.md kind {value!r} is not one of [{allowed}]: {manifest_path}"
        )
    return value


def _parse_tools(
    data: dict[str, Any],
    plugin_dir: Path,
    manifest_path: Path,
    *,
    plugin_name: str,
) -> list[ToolEntry]:
    tools_raw = data.get("tools")
    if not isinstance(tools_raw, list) or not tools_raw:
        raise PluginManifestError(
            f"plugin.md tools must be a non-empty list: {manifest_path}"
        )
    expected_prefix = f"{plugin_name}."
    seen_names: set[str] = set()
    tools: list[ToolEntry] = []
    for index, entry in enumerate(tools_raw):
        if not isinstance(entry, dict):
            raise PluginManifestError(
                f"plugin.md tools[{index}] must be a mapping: {manifest_path}"
            )
        tool_name = _require_str(
            entry, "name", manifest_path, path=f"tools[{index}].name"
        )
        if not tool_name.startswith(expected_prefix):
            raise PluginManifestError(
                f"plugin.md tools[{index}].name {tool_name!r} must start "
                f"with {expected_prefix!r}: {manifest_path}"
            )
        if tool_name in seen_names:
            raise PluginManifestError(
                f"plugin.md duplicate tool name {tool_name!r}: {manifest_path}"
            )
        seen_names.add(tool_name)
        module_raw = _require_str(
            entry, "module", manifest_path, path=f"tools[{index}].module"
        )
        module_path = _resolve_under(
            plugin_dir,
            module_raw,
            manifest_path,
            field=f"tools[{index}].module",
        )
        if not module_path.is_file():
            raise PluginManifestError(
                f"plugin.md tools[{index}].module path does not exist: "
                f"{module_path}"
            )
        tools.append(ToolEntry(name=tool_name, module=module_path))
    return tools


def _resolve_setup(
    data: dict[str, Any],
    plugin_dir: Path,
    manifest_path: Path,
) -> Path | None:
    raw = data.get("setup")
    if raw is None:
        default = plugin_dir / "setup.sh"
        return default if default.is_file() else None
    if not isinstance(raw, str) or not raw.strip():
        raise PluginManifestError(
            "plugin.md setup must be a non-empty string when set: "
            f"{manifest_path}"
        )
    setup_path = _resolve_under(
        plugin_dir, raw.strip(), manifest_path, field="setup"
    )
    if not setup_path.is_file():
        raise PluginManifestError(
            f"plugin.md setup path does not exist: {setup_path}"
        )
    return setup_path


def _resolve_optional_path(
    data: dict[str, Any],
    field: str,
    plugin_dir: Path,
    manifest_path: Path,
) -> Path | None:
    raw = data.get(field)
    if raw is None:
        return None
    if not isinstance(raw, str) or not raw.strip():
        raise PluginManifestError(
            f"plugin.md {field} must be a non-empty string when set: "
            f"{manifest_path}"
        )
    candidate = _resolve_under(
        plugin_dir, raw.strip(), manifest_path, field=field
    )
    if not candidate.is_file():
        raise PluginManifestError(
            f"plugin.md {field} path does not exist: {candidate}"
        )
    return candidate


def _require_str(
    data: dict[str, Any],
    field: str,
    manifest_path: Path,
    *,
    path: str | None = None,
) -> str:
    value = data.get(field)
    if not isinstance(value, str) or not value.strip():
        loc = path or field
        raise PluginManifestError(
            f"plugin.md {loc} must be a non-empty string: {manifest_path}"
        )
    return value.strip()


def _resolve_under(
    plugin_dir: Path,
    raw: str,
    manifest_path: Path,
    *,
    field: str,
) -> Path:
    candidate = (plugin_dir / raw).resolve()
    try:
        candidate.relative_to(plugin_dir)
    except ValueError as exc:
        raise PluginManifestError(
            f"plugin.md {field} path escapes plugin dir: {raw!r}: "
            f"{manifest_path}"
        ) from exc
    return candidate

File: agent-core/crates/eos-plugin-catalog/src/manifest.rs
//! [`PluginManifest`] parsing: the two-stage parse from `plugin.md` frontmatter.
//!
//! Ports `plugins/core/manifest.py`. Validation is two-stage (`api-parse-dont-validate`):
//! deserialize the frontmatter into a tolerant [`RawManifest`] DTO (the only
//! `Deserialize` target here), then **validate-into** the invariant-bearing
//! [`PluginManifest`] with `plugin_dir` context. `RawManifest`'s fields are
//! `Option<serde_yaml::Value>` rather than `Option<String>` so that a wrong-typed
//! field (e.g. `name: 123`) yields the granular [`PluginCatalogError::MissingField`]
//! / [`PluginCatalogError::KindNotString`] that Python's `isinstance`-style
//! `_require_str`/`_parse_kind` produce, instead of an opaque serde error — the
//! source is the tie-breaker over the spec's `Option<String>` sketch.

use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::error::PluginCatalogError;
use crate::frontmatter::split_frontmatter;
use crate::names::{PluginName, PluginResolvedPath, PluginToolName};

/// The declared `kind` of a plugin (was `ALLOWED_PLUGIN_KINDS`, manifest.py
/// 45-54). `#[non_exhaustive]` — the plan reserves room for new kinds.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
#[non_exhaustive]
pub enum PluginKind {
    /// A language server (e.g. the bundled LSP plugin).
    LanguageServer,
    /// A code formatter.
    Formatter,
    /// A source indexer.
    Indexer,
    /// A long-lived build daemon.
    BuildDaemon,
    /// A bridge to an external MCP server.
    McpBridge,
    /// An uncategorized plugin (the audit fallback when `kind` is unset).
    Custom,
}

impl PluginKind {
    /// Parse a recognized kind string, returning
    /// [`PluginCatalogError::UnknownKind`] for an unrecognized value.
    pub(crate) fn parse(s: &str) -> Result<Self, PluginCatalogError> {
        match s {
            "language_server" => Ok(Self::LanguageServer),
            "formatter" => Ok(Self::Formatter),
            "indexer" => Ok(Self::Indexer),
            "build_daemon" => Ok(Self::BuildDaemon),
            "mcp_bridge" => Ok(Self::McpBridge),
            "custom" => Ok(Self::Custom),
            other => Err(PluginCatalogError::UnknownKind(other.to_owned())),
        }
    }
}

/// One declared tool in a manifest (manifest.py 57-62).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, JsonSchema)]
#[non_exhaustive]
pub struct ToolEntry {
    /// The validated `<plugin_name>.<suffix>` tool name.
    pub name: PluginToolName,
    /// The tool module path, resolved under the plugin directory and proven to
    /// exist on disk (validated only — **never** imported or executed).
    pub module: PluginResolvedPath,
}

/// A parsed and validated `plugin.md` (manifest.py 65-77). An immutable value
/// type produced **only** by [`parse_plugin_manifest`] — it derives no
/// `Deserialize` (it is not a wire-input DTO; see [`RawManifest`]).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, JsonSchema)]
#[non_exhaustive]
pub struct PluginManifest {
    /// The plugin name (equals the plugin directory name).
    pub name: PluginName,
    /// A human-readable description.
    pub description: String,
    /// The non-empty list of declared tools.
    pub tools: Vec<ToolEntry>,
    /// The optional setup script path (validated, not executed).
    pub setup: Option<PluginResolvedPath>,
    /// The optional in-sandbox runtime entrypoint path (validated, not executed).
    pub runtime: Option<PluginResolvedPath>,
    /// The absolute, canonicalized plugin directory.
    pub source_dir: PathBuf,
    /// The trimmed markdown body after the frontmatter (informational).
    pub body: String,
    /// The optional declared kind (`None` when unset).
    pub kind: Option<PluginKind>,
}

/// Tolerant wire-input DTO: the only `Deserialize` target in this crate. Each
/// field is `Option<serde_yaml::Value>` so a wrong-typed field surfaces as a
/// granular [`PluginCatalogError`] during validation rather than collapsing into
/// one opaque serde error. Unknown frontmatter keys are ignored (no
/// `deny_unknown_fields`), matching Python's `dict.get`.
#[derive(Debug, Deserialize)]
pub(crate) struct RawManifest {
    #[serde(default)]
    name: Option<serde_yaml::Value>,
    #[serde(default)]
    description: Option<serde_yaml::Value>,
    #[serde(default)]
    tools: Option<serde_yaml::Value>,
    #[serde(default)]
    setup: Option<serde_yaml::Value>,
    #[serde(default)]
    runtime: Option<serde_yaml::Value>,
    #[serde(default)]
    kind: Option<serde_yaml::Value>,
}

/// Parse `<plugin_dir>/plugin.md` and validate its schema.
///
/// # Errors
/// A [`PluginCatalogError`] variant for each distinct failure (missing manifest,
/// bad frontmatter, name/dir mismatch, tool prefix/duplicate, path escape/missing,
/// kind validation), mirroring `parse_plugin_manifest` in manifest.py.
pub(crate) fn parse_plugin_manifest(
    plugin_dir: &Path,
) -> Result<PluginManifest, PluginCatalogError> {
    let plugin_dir = plugin_dir
        .canonicalize()
        .map_err(|cause| PluginCatalogError::Io {
            path: plugin_dir.to_owned(),
            cause,
        })?;
    let manifest_path = plugin_dir.join("plugin.md");
    if !manifest_path.is_file() {
        return Err(PluginCatalogError::ManifestMissing(plugin_dir));
    }
    let text = std::fs::read_to_string(&manifest_path).map_err(|cause| PluginCatalogError::Io {
        path: manifest_path.clone(),
        cause,
    })?;
    let (frontmatter, body) = split_frontmatter(&text)
        .ok_or_else(|| PluginCatalogError::MissingFrontmatter(manifest_path.clone()))?;

    let value: serde_yaml::Value =
        serde_yaml::from_str(&frontmatter).map_err(|cause| PluginCatalogError::Frontmatter {
            path: manifest_path.clone(),
            cause,
        })?;
    // `yaml.safe_load(...) or {}`: an empty frontmatter is an empty mapping.
    let value = if value.is_null() {
        serde_yaml::Value::Mapping(serde_yaml::Mapping::new())
    } else {
        value
    };
    if !value.is_mapping() {
        return Err(PluginCatalogError::NotMapping(manifest_path));
    }
    let raw: RawManifest =
        serde_yaml::from_value(value).map_err(|cause| PluginCatalogError::Frontmatter {
            path: manifest_path.clone(),
            cause,
        })?;

    let dir_name = plugin_dir
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or_default()
        .to_owned();

    let name = require_str(raw.name.as_ref(), "name", &manifest_path)?;
    let plugin_name = PluginName::parse(name.clone())?;
    if name != dir_name {
        return Err(PluginCatalogError::NameDirMismatch {
            name,
            dir: dir_name,
        });
    }

    let description = require_str(raw.description.as_ref(), "description", &manifest_path)?;
    let tools = parse_tools(raw.tools.as_ref(), &plugin_dir, &manifest_path, &name)?;
    let setup = resolve_setup(raw.setup.as_ref(), &plugin_dir, &manifest_path)?;
    let runtime =
        resolve_optional_path(raw.runtime.as_ref(), "runtime", &plugin_dir, &manifest_path)?;
    let kind = parse_kind(raw.kind.as_ref(), &manifest_path)?;

    Ok(PluginManifest {
        name: plugin_name,
        description,
        tools,
        setup,
        runtime,
        source_dir: plugin_dir,
        body: body.trim().to_owned(),
        kind,
    })
}

/// Extract a required non-empty string field, else [`PluginCatalogError::MissingField`].
fn require_str(
    field: Option<&serde_yaml::Value>,
    name: &str,
    path: &Path,
) -> Result<String, PluginCatalogError> {
    string_value(field).ok_or_else(|| PluginCatalogError::MissingField {
        path: path.to_owned(),
        field: name.to_owned(),
    })
}

/// `Some(trimmed)` iff the value is a non-empty string (Python `isinstance(str)`
/// + `.strip()`), else `None`.
fn string_value(v: Option<&serde_yaml::Value>) -> Option<String> {
    match v {
        Some(serde_yaml::Value::String(s)) if !s.trim().is_empty() => Some(s.trim().to_owned()),
        _ => None,
    }
}

fn parse_tools(
    tools: Option<&serde_yaml::Value>,
    plugin_dir: &Path,
    manifest_path: &Path,
    plugin_name: &str,
) -> Result<Vec<ToolEntry>, PluginCatalogError> {
    let seq = match tools {
        Some(serde_yaml::Value::Sequence(seq)) if !seq.is_empty() => seq,
        _ => return Err(PluginCatalogError::EmptyTools(manifest_path.to_owned())),
    };
    let prefix = format!("{plugin_name}.");
    let mut seen: BTreeSet<String> = BTreeSet::new();
    let mut entries = Vec::with_capacity(seq.len());
    for (index, entry) in seq.iter().enumerate() {
        let map = entry.as_mapping();
        let tool_name = string_value(map.and_then(|m| m.get("name"))).ok_or_else(|| {
            PluginCatalogError::MissingField {
                path: manifest_path.to_owned(),
                field: format!("tools[{index}].name"),
            }
        })?;
        if !tool_name.starts_with(&prefix) {
            return Err(PluginCatalogError::ToolPrefix {
                name: tool_name,
                prefix,
            });
        }
        if !seen.insert(tool_name.clone()) {
            return Err(PluginCatalogError::DuplicateTool(tool_name));
        }
        let module_raw = string_value(map.and_then(|m| m.get("module"))).ok_or_else(|| {
            PluginCatalogError::MissingField {
                path: manifest_path.to_owned(),
                field: format!("tools[{index}].module"),
            }
        })?;
        let module = PluginResolvedPath::resolve_under(plugin_dir, &module_raw)?;
        if !module.as_path().is_file() {
            return Err(PluginCatalogError::PathMissing(module.into_path_buf()));
        }
        entries.push(ToolEntry {
            name: PluginToolName::new(tool_name),
            module,
        });
    }
    Ok(entries)
}

/// Resolve `setup`, defaulting to `setup.sh` iff it exists when unset
/// (manifest.py 218-239).
fn resolve_setup(
    raw: Option<&serde_yaml::Value>,
    plugin_dir: &Path,
    manifest_path: &Path,
) -> Result<Option<PluginResolvedPath>, PluginCatalogError> {
    // The unset case defaults to `setup.sh` iff present; the set-string and error
    // arms are identical to `resolve_optional_path`, so delegate them (DRY).
    if matches!(raw, None | Some(serde_yaml::Value::Null)) {
        let default = PluginResolvedPath::resolve_under(plugin_dir, "setup.sh")?;
        return Ok(default.as_path().is_file().then_some(default));
    }
    resolve_optional_path(raw, "setup", plugin_dir, manifest_path)
}

/// Resolve an optional declared path (`runtime`); `None` when unset, error when
/// set-but-not-a-non-empty-string (manifest.py 242-263).
fn resolve_optional_path(
    raw: Option<&serde_yaml::Value>,
    field: &str,
    plugin_dir: &Path,
    manifest_path: &Path,
) -> Result<Option<PluginResolvedPath>, PluginCatalogError> {
    match raw {
        None | Some(serde_yaml::Value::Null) => Ok(None),
        Some(serde_yaml::Value::String(s)) if !s.trim().is_empty() => {
            resolve_existing(plugin_dir, s.trim())
        }
        _ => Err(PluginCatalogError::MissingField {
            path: manifest_path.to_owned(),
            field: field.to_owned(),
        }),
    }
}

/// Resolve a declared path under the plugin dir and require it to exist.
fn resolve_existing(
    plugin_dir: &Path,
    raw: &str,
) -> Result<Option<PluginResolvedPath>, PluginCatalogError> {
    let resolved = PluginResolvedPath::resolve_under(plugin_dir, raw)?;
    if resolved.as_path().is_file() {
        Ok(Some(resolved))
    } else {
        Err(PluginCatalogError::PathMissing(resolved.into_path_buf()))
    }
}

/// Validate the optional `kind` field (manifest.py 140-164).
fn parse_kind(
    raw: Option<&serde_yaml::Value>,
    manifest_path: &Path,
) -> Result<Option<PluginKind>, PluginCatalogError> {
    match raw {
        None | Some(serde_yaml::Value::Null) => Ok(None),
        Some(serde_yaml::Value::String(s)) if !s.trim().is_empty() => {
            PluginKind::parse(s.trim()).map(Some)
        }
        _ => Err(PluginCatalogError::KindNotString(manifest_path.to_owned())),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::support::{make_plugin, temp_root};

    const LSP_MANIFEST: &str = "\
---
name: lsp
description: Pyright-backed LSP tools for Python.
kind: language_server
tools:
  - name: lsp.hover
    module: tools/hover.py
  - name: lsp.find_definitions
    module: tools/find_definitions.py
  - name: lsp.find_references
    module: tools/find_references.py
  - name: lsp.diagnostics
    module: tools/diagnostics.py
  - name: lsp.query_symbols
    module: tools/query_symbols.py
  - name: lsp.apply_workspace_edit
    module: tools/apply_workspace_edit.py
  - name: lsp.rename
    module: tools/rename.py
  - name: lsp.format
    module: tools/format.py
  - name: lsp.code_actions
    module: tools/code_actions.py
  - name: lsp.apply_code_action
    module: tools/apply_code_action.py
setup: setup.sh
runtime: runtime/server.py
---

# LSP Plugin

Pyright-backed Python language tools.
";

    const LSP_FILES: &[&str] = &[
        "tools/hover.py",
        "tools/find_definitions.py",
        "tools/find_references.py",
        "tools/diagnostics.py",
        "tools/query_symbols.py",
        "tools/apply_workspace_edit.py",
        "tools/rename.py",
        "tools/format.py",
        "tools/code_actions.py",
        "tools/apply_code_action.py",
        "setup.sh",
        "runtime/server.py",
    ];

    // AC-plugin-catalog-01: the real LSP manifest parses to the expected shape.
    #[test]
    fn parses_lsp_manifest() {
        let root = temp_root("lsp_ok");
        let dir = make_plugin(&root, "lsp", Some(LSP_MANIFEST), LSP_FILES);
        let manifest = parse_plugin_manifest(&dir).expect("parses");
        assert_eq!(manifest.name.as_str(), "lsp");
        assert_eq!(manifest.kind, Some(PluginKind::LanguageServer));
        assert_eq!(manifest.tools.len(), 10);
        for tool in &manifest.tools {
            assert!(tool.name.as_str().starts_with("lsp."));
            assert!(tool.module.as_path().is_file());
        }
        assert!(manifest.setup.is_some());
        assert!(manifest.runtime.is_some());
        assert!(manifest.body.contains("LSP Plugin"));
        let _ = std::fs::remove_dir_all(&root);
    }

    // AC-plugin-catalog-02: the manifest rejection paths.
    #[test]
    fn rejects_bad_names_prefixes_and_duplicates() {
        let root = temp_root("reject");

        let d1 = make_plugin(
            &root,
            "lsp",
            Some("---\nname: other\ndescription: d\ntools:\n  - name: other.x\n    module: tools/x.py\n---\n"),
            &[],
        );
        assert!(matches!(
            parse_plugin_manifest(&d1),
            Err(PluginCatalogError::NameDirMismatch { .. })
        ));

        let d2 = make_plugin(
            &root,
            "alpha",
            Some("---\nname: alpha\ndescription: d\ntools:\n  - name: beta.x\n    module: tools/x.py\n---\n"),
            &[],
        );
        assert!(matches!(
            parse_plugin_manifest(&d2),
            Err(PluginCatalogError::ToolPrefix { .. })
        ));

        // The first tool's module must exist so the loop reaches the duplicate
        // before any per-tool module-existence check fires (parse order matches
        // Python's `_parse_tools`).
        let d3 = make_plugin(
            &root,
            "gamma",
            Some("---\nname: gamma\ndescription: d\ntools:\n  - name: gamma.x\n    module: a.py\n  - name: gamma.x\n    module: b.py\n---\n"),
            &["a.py"],
        );
        assert!(matches!(
            parse_plugin_manifest(&d3),
            Err(PluginCatalogError::DuplicateTool(_))
        ));

        let d4 = make_plugin(&root, "seqp", Some("---\n- a\n- b\n---\n"), &[]);
        assert!(matches!(
            parse_plugin_manifest(&d4),
            Err(PluginCatalogError::NotMapping(_))
        ));

        let d5 = make_plugin(
            &root,
            "empty",
            Some("---\nname: empty\ndescription: d\ntools: []\n---\n"),
            &[],
        );
        assert!(matches!(
            parse_plugin_manifest(&d5),
            Err(PluginCatalogError::EmptyTools(_))
        ));

        let d6 = make_plugin(
            &root,
            "noname",
            Some("---\ndescription: d\ntools:\n  - name: noname.x\n    module: tools/x.py\n---\n"),
            &[],
        );
        assert!(matches!(
            parse_plugin_manifest(&d6),
            Err(PluginCatalogError::MissingField { .. })
        ));

        // No `---` fence at all -> MissingFrontmatter (manifest.py 87-92).
        let d7 = make_plugin(&root, "nofence", Some("name: nofence\n"), &[]);
        assert!(matches!(
            parse_plugin_manifest(&d7),
            Err(PluginCatalogError::MissingFrontmatter(_))
        ));

        // Fenced but malformed YAML -> Frontmatter (manifest.py 96-99).
        let d8 = make_plugin(&root, "badyaml", Some("---\nname: [unterminated\n---\n"), &[]);
        assert!(matches!(
            parse_plugin_manifest(&d8),
            Err(PluginCatalogError::Frontmatter { .. })
        ));

        let _ = std::fs::remove_dir_all(&root);
    }

    // AC-plugin-catalog-03: a declared path that is under-dir but does not exist.
    #[test]
    fn missing_declared_path_errors() {
        let root = temp_root("missing_path");
        let dir = make_plugin(
            &root,
            "alpha",
            Some("---\nname: alpha\ndescription: d\ntools:\n  - name: alpha.x\n    module: tools/x.py\n---\n"),
            &[],
        );
        assert!(matches!(
            parse_plugin_manifest(&dir),
            Err(PluginCatalogError::PathMissing(_))
        ));
        let _ = std::fs::remove_dir_all(&root);
    }

    // AC-plugin-catalog-05: kind validation (unset -> None; known; unknown -> error).
    #[test]
    fn kind_enum_validation() {
        let root = temp_root("kind");
        let manifest = |kind_line: &str| {
            format!(
                "---\nname: k\ndescription: d\n{kind_line}tools:\n  - name: k.x\n    module: tools/x.py\n---\n"
            )
        };

        let d_unset = make_plugin(&root, "k", Some(&manifest("")), &["tools/x.py"]);
        assert_eq!(parse_plugin_manifest(&d_unset).expect("parses").kind, None);

        let d_known = make_plugin(&root, "k", Some(&manifest("kind: formatter\n")), &["tools/x.py"]);
        assert_eq!(
            parse_plugin_manifest(&d_known).expect("parses").kind,
            Some(PluginKind::Formatter)
        );

        let d_unknown = make_plugin(&root, "k", Some(&manifest("kind: wizard\n")), &["tools/x.py"]);
        assert!(matches!(
            parse_plugin_manifest(&d_unknown),
            Err(PluginCatalogError::UnknownKind(_))
        ));

        // A present-but-non-string kind -> KindNotString (manifest.py 154-157).
        let d_nonstr = make_plugin(&root, "k", Some(&manifest("kind: 123\n")), &["tools/x.py"]);
        assert!(matches!(
            parse_plugin_manifest(&d_nonstr),
            Err(PluginCatalogError::KindNotString(_))
        ));

        let _ = std::fs::remove_dir_all(&root);
    }

    // AC-plugin-catalog-01 (setup default branch): an unset `setup:` resolves to
    // `setup.sh` when that file exists on disk (manifest.py 224-226).
    #[test]
    fn setup_defaults_to_setup_sh_when_present() {
        let root = temp_root("setup_default");
        let dir = make_plugin(
            &root,
            "s",
            Some("---\nname: s\ndescription: d\ntools:\n  - name: s.x\n    module: tools/x.py\n---\n"),
            &["tools/x.py", "setup.sh"],
        );
        let manifest = parse_plugin_manifest(&dir).expect("parses");
        let setup = manifest.setup.expect("setup defaults to setup.sh");
        assert!(setup.as_path().ends_with("setup.sh"));
        let _ = std::fs::remove_dir_all(&root);
    }
}
